docs/contributing/codebase-overview.md

```diff
@@ -0,0 +1,262 @@
+---
+id: codebase-overview
+title: Codebase Overview
+layout: contributing
+permalink: contributing/codebase-overview.html
+prev: how-to-contribute.html
+next: design-principles.html
+---
+
+This section will give you an overview of the React codebase organization, its conventions, and the implementation.
+
+If you want to [contribute to React](/react/contributing/how-to-contribute.html) we hope that this guide will help you feel more comfortable making changes.
+
+We don't necessarily recommend any of these conventions in React apps. Many of them exist for historical reasons and might change with time.
+
+### Custom Module System
+
+At Facebook, internally we use a custom module system called "Haste". It is similar to [CommonJS](https://nodejs.org/docs/latest/api/modules.html) and also uses `require()` but has a few important differences that often confuse outside contributors.
+
+In CommonJS, when you import a module, you need to specify its relative path:
+
+```js
+// Importing from the same folder:
+var setInnerHTML = require('./setInnerHTML');
+
+// Importing from a different folder:
+var setInnerHTML = require('../utils/setInnerHTML');
+
+// Importing from a deeply nested folder:
+var setInnerHTML = require('../client/utils/setInnerHTML');
+```
+
+However, with Haste **all filenames are globally unique.** In the React codebase, you can import any module from any other module by its name alone:
+
+```js
+var setInnerHTML = require('setInnerHTML');
+```
+
+Haste was originally developed for giant apps like Facebook. It's easy to move files to different folders and import them without worrying about relative paths. The fuzzy file search in any editor always takes you to the correct place thanks to globally unique names.
```

Comment:
Technically, haste doesn't have the constraint that filename needs to be unique nor that it matches the name that you require. Haste only looks at the `@providesModule` token in the first comment of the file.

Most people on fb codebase use the same filename because why invent a new name and it indeeds make it easier to search files by name.

Now, React in addition to haste has an export step to npm that takes all the files, move them to the root directory and rewrites all the require('Module') to require('./Module') which uses the filename.

So in the React codebase, filenames must match the name of the @providesModule. React Native doesn't have the same step so you can name your files whatever you want for example.

Feel free to leave this comment as is, but I figured I would let you know what is exactly happening :)